File: code/typescript-for-csharp-java/normalize.mjs
// Makes tsc, Node.js, dotnet and javac output comparable between machines: node normalize.mjs < raw.txt > out.txt
// - the ANSI colors of tsc --pretty are removed
// - absolute paths and file URLs under this folder become paths relative to it, with forward slashes
// - stack frames ("    at …") are dropped: they name Node's internal files and lines (a last frame followed by
//   the error's properties keeps its opening brace)
// - the process ID of a warning becomes PID
import { readFileSync } from 'node:fs';

const course = 'typescript-for-csharp-java';
const text = readFileSync(0, 'utf8').replaceAll('\r\n', '\n');
const pathPattern = new RegExp(String.raw`(?:file://)?(?:\\\\\?\\)?[^\s'"(]*?${course}(?:[\\/]([^\s'"(:]*))?`, 'g');

const lines = text
  .replace(/\x1b\[[0-9;]*m/g, '')
  .split('\n')
  .map((line) => (/^\s+at .* \{$/.test(line) ? '{' : line))
  .filter((line) => !/^\s+at /.test(line))
  .map((line) => line.replace(pathPattern, (_, rest) => (rest ?? '.').replaceAll('\\', '/')))
  .map((line) => line.replace(/^\(node:\d+\)/, '(node:PID)'));

process.stdout.write(lines.join('\n'));

File: code/python-for-csharp-java/normalize.py
# Makes the outputs of python, mypy, pytest, uv, dotnet and javac comparable between machines:
#     python normalize.py [--uv] < raw.txt > out.txt
# - colors are removed, and line endings become \n
# - absolute paths and file URLs under this folder become paths relative to it, with forward slashes,
#   and so do relative paths that Windows prints with backslashes (errors\l01_order.py)
# - memory addresses (<function f at 0x7f3a…>) become 0xADDRESS, and pytest's durations become TIME
# - with --uv, uv's progress lines (Resolved…, Installed…, + package==version) are dropped: they depend on its cache
import re
import sys

COURSE = "python-for-csharp-java"

course_path = re.compile(r"(?:file:///?)?[^\s'\"(`]*?" + re.escape(COURSE) + r"(?:[\\/]([^\s'\":(),`]*))?")
relative_windows_path = re.compile(r"\b(?:[\w.-]+\\)+[\w.-]+\.(?:py|cs|java|toml|json)\b")
uv_progress = re.compile(
    r"^\s*(?:Resolved|Prepared|Installed|Uninstalled|Audited|Downloading|Downloaded|Building|Built|Using CPython"
    r"|Creating virtual environment|Removed virtual environment|Bytecode compiled)\b|^ [+~-] \S+(?:==| @ )"
)


def normalize_line(line: str) -> str:
    line = re.sub(r"\x1b\[[0-9;]*[A-Za-z]", "", line)
    line = course_path.sub(lambda m: (m.group(1) or ".").replace("\\", "/"), line)
    line = relative_windows_path.sub(lambda m: m.group(0).replace("\\", "/"), line)
    line = re.sub(r"\b0x[0-9A-Fa-f]{6,16}\b", "0xADDRESS", line)
    line = re.sub(r" in \d+\.\d+s\b", " in TIME", line)
    return line.rstrip()


def main() -> None:
    drop_uv = "--uv" in sys.argv[1:]
    text = sys.stdin.buffer.read().decode("utf-8", errors="replace").replace("\r\n", "\n")
    lines = [normalize_line(line) for line in text.split("\n")]
    if drop_uv:
        lines = [line for line in lines if not uv_progress.search(line)]
    sys.stdout.buffer.write("\n".join(lines).encode("utf-8"))


main()
